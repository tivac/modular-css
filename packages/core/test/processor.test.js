"use strict";

var path    = require("path"),
    assert  = require("assert"),

    leading = require("dentist").dedent,
    
    Processor = require("../processor.js"),
    
    compare = require("test-utils/compare.js")(__dirname);

describe("/processor.js", function() {
    describe("Basics", function() {
        it("should be a function", function() {
            expect(typeof Processor).toBe("function");
        });
        
        it("should auto-instantiate if called without new", function() {
            /* eslint new-cap:0 */
            assert(Processor() instanceof Processor);
        });
    });

    describe("functionality", function() {
        beforeEach(function() {
            this.processor = new Processor({
                namer : (file, selector) => selector
            });
        });
        
        describe("getters", function() {
            describe(".file", function() {
                it("should return all the files that have been added", function() {
                    var processor = this.processor;
                    
                    return processor.file(
                        "./packages/core/test/specimens/start.css"
                    )
                    .then(() => processor.file("./packages/core/test/specimens/local.css"))
                    .then(() => {
                        expect(typeof processor.files).toBe("object");
                        
                        expect(Object.keys(processor.files).length).toBe(3);
                        
                        assert(path.resolve("./packages/core/test/specimens/local.css") in processor.files);
                        assert(path.resolve("./packages/core/test/specimens/start.css") in processor.files);
                        assert(path.resolve("./packages/core/test/specimens/folder/folder.css") in processor.files);
                    });
                });
            });
        });
        
        describe("bad imports", function() {
            it("should fail if a value imports a non-existant reference", function() {
                return this.processor.string(
                    "./invalid/value.css",
                    "@value not-real from \"../local.css\";"
                )
                .catch((error) => expect(error.message).toBe(
                    `Unable to locate "../local.css" from "${path.resolve("./invalid/value.css")}"`
                ));
            });
            
            it("should fail if a composition imports a non-existant reference", function() {
                return this.processor.string(
                    "./invalid/composition.css",
                    ".wooga { composes: fake from \"../local.css\"; }"
                )
                .catch((error) => expect(error.message).toBe(
                    `Unable to locate "../local.css" from "${path.resolve("./invalid/composition.css")}"`
                ));
            });
        });

        describe("scoping", function() {
            it("should scope classes, ids, and keyframes", function() {
                return this.processor.string(
                    "./packages/core/test/specimens/simple.css",
                    leading(`
                        @keyframes kooga { }
                        #fooga { }
                        .wooga { }
                        .one,
                        .two { }
                    `)
                )
                .then((result) => {
                    expect(result.exports).toEqual({
                        fooga : [ "fooga" ],
                        wooga : [ "wooga" ],
                        one   : [ "one" ],
                        two   : [ "two" ]
                    });

                    return this.processor.output();
                })
                .then((output) =>
                    expect(output.css).toEqual(
                        leading(`
                            /* packages/core/test/specimens/simple.css */
                            @keyframes kooga {}
                            #fooga {}
                            .wooga {}
                            .one,
                            .two {}
                        `)
                    )
                );
            });
        });

        describe("values", function() {
            it("should support local values in value composition", function() {
                return this.processor.string(
                    "./packages/core/test/specimens/simple.css",
                    "@value local: './local.css'; @value one from local; .fooga { background: one; }"
                )
                .then((result) => expect(result.exports).toEqual({
                    fooga : [ "fooga" ]
                }));
            });
        });

        describe("externals", function() {
            it("should support overriding external values", function() {
                return this.processor.file(
                    "./packages/core/test/specimens/externals.css"
                )
                .then(() => this.processor.output())
                .then((result) => compare.stringToFile(result.css, "./packages/core/test/results/processor/externals.css"));
            });
        });

        describe("exports", function() {
            it("should export an object of arrays containing strings", function() {
                return this.processor.string(
                    "./packages/core/test/specimens/simple.css",
                    ".red { color: red; } .black { background: #000; } .one, .two { composes: red, black; }"
                )
                .then((result) => expect(result.exports).toEqual({
                    red   : [ "red" ],
                    black : [ "black" ],
                    one   : [ "red", "black", "one" ],
                    two   : [ "red", "black", "two" ]
                }))
                .catch((e) => {
                    throw e;
                });
            });

            it("should export identifiers and their classes", function() {
                return this.processor.file(
                    "./packages/core/test/specimens/start.css"
                )
                .then(() => this.processor.output())
                .then((output) => expect(output.compositions).toMatchSnapshot());
            });
        });

        it("should support unicode classes & ids", function() {
            var processor = new Processor({
                    namer : (file, selector) => selector
                });
            
            return processor.file(
                "./packages/core/test/specimens/processor/unicode.css"
            )
            .then(() => processor.output({ to : "./packages/core/test/output/processor/unicode.css" }))
            .then((output) => compare.stringToFile(output.css, "./packages/core/test/results/processor/unicode.css"));
        });
    });
});
