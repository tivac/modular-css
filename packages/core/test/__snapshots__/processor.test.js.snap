// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/processor.js functionality exports should export an object of arrays containing strings 1`] = `
Object {
  "black": Array [
    "black",
  ],
  "one": Array [
    "red",
    "black",
    "one",
  ],
  "red": Array [
    "red",
  ],
  "two": Array [
    "red",
    "black",
    "two",
  ],
}
`;

exports[`/processor.js functionality exports should export identifiers and their classes 1`] = `
Object {
  "packages/core/test/specimens/folder/folder.css": Object {
    "folder": "folder",
  },
  "packages/core/test/specimens/local.css": Object {
    "booga": "booga",
    "looga": "booga looga",
  },
  "packages/core/test/specimens/start.css": Object {
    "booga": "booga",
    "tooga": "tooga",
    "wooga": "booga wooga",
  },
}
`;

exports[`/processor.js functionality externals should support overriding external values 1`] = `
"/* packages/core/test/specimens/folder/folder.css */
.folder {
    margin: 2px
}
/* packages/core/test/specimens/simple.css */
.wooga {
    color: red
}
/* packages/core/test/specimens/local.css */
.booga {
    background: green
}
/* packages/core/test/specimens/externals.css */
.a .wooga {
    color: green
}
.b .booga {
    background: blue
}"
`;

exports[`/processor.js functionality getters .file should return all the files that have been added 1`] = `
Array [
  "packages/core/test/specimens/start.css",
  "packages/core/test/specimens/local.css",
  "packages/core/test/specimens/folder/folder.css",
]
`;

exports[`/processor.js functionality scoping should scope classes, ids, and keyframes 1`] = `
Object {
  "fooga": Array [
    "fooga",
  ],
  "one": Array [
    "one",
  ],
  "two": Array [
    "two",
  ],
  "wooga": Array [
    "wooga",
  ],
}
`;

exports[`/processor.js functionality scoping should scope classes, ids, and keyframes 2`] = `
"/* simple.css */
@keyframes kooga {}
#fooga {}
.wooga {}
.one,
.two {}"
`;

exports[`/processor.js functionality should support unicode classes & ids 1`] = `
"/* packages/core/test/specimens/processor/unicode.css */
#‚ô• {
    background: lime
}
#¬© {
    background: lime
}
#‚Äú‚Äò‚Äô‚Äù {
    background: lime
}
#‚ò∫‚òÉ {
    background: lime
}
#‚åò‚å• {
    background: lime
}
#ùÑû‚ô™‚ô©‚ô´‚ô¨ {
    background: lime
}
#üí© {
    background: lime
}
#\\\\? {
    background: lime
}
#\\\\@ {
    background: lime
}
#\\\\. {
    background: lime
}
#\\\\3A \\\\) {
    background: lime
}
#\\\\3A \\\\\`\\\\( {
    background: lime
}
#\\\\31 23 {
    background: lime
}
#\\\\31 a2b3c {
    background: lime
}
#\\\\<p\\\\> {
    background: lime
}
#\\\\<\\\\>\\\\<\\\\<\\\\<\\\\>\\\\>\\\\<\\\\> {
    background: lime
}
#\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\[\\\\>\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\>\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\>\\\\+\\\\+\\\\+\\\\>\\\\+\\\\<\\\\<\\\\<\\\\<\\\\-\\\\]\\\\>\\\\+\\\\+\\\\.\\\\>\\\\+\\\\.\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\.\\\\.\\\\+\\\\+\\\\+\\\\.\\\\>\\\\+\\\\+\\\\.\\\\<\\\\<\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\.\\\\>\\\\.\\\\+\\\\+\\\\+\\\\.\\\\-\\\\-\\\\-\\\\-\\\\-\\\\-\\\\.\\\\-\\\\-\\\\-\\\\-\\\\-\\\\-\\\\-\\\\-\\\\.\\\\>\\\\+\\\\.\\\\>\\\\. {
    background: lime
}
#\\\\# {
    background: lime
}
#\\\\#\\\\# {
    background: lime
}
#\\\\#\\\\.\\\\#\\\\.\\\\# {
    background: lime
}
#\\\\_ {
    background: lime
}
#\\\\{\\\\} {
    background: lime
}
#\\\\.fake\\\\-class {
    background: lime
}
#foo\\\\.bar {
    background: lime
}
#\\\\3A hover {
    background: lime
}
#\\\\3A hover\\\\3A focus\\\\3A active {
    background: lime
}
#\\\\[attr\\\\=value\\\\] {
    background: lime
}
#f\\\\/o\\\\/o {
    background: lime
}
#f\\\\\\\\o\\\\\\\\o {
    background: lime
}
#f\\\\*o\\\\*o {
    background: lime
}
#f\\\\!o\\\\!o {
    background: lime
}
#f\\\\'o\\\\'o {
    background: lime
}
#f\\\\~o\\\\~o {
    background: lime
}
#f\\\\+o\\\\+o {
    background: lime
}
.‚ô• {
    background: lime
}
.¬© {
    background: lime
}
.‚Äú‚Äò‚Äô‚Äù {
    background: lime
}
.‚ò∫‚òÉ {
    background: lime
}
.‚åò‚å• {
    background: lime
}
.ùÑû‚ô™‚ô©‚ô´‚ô¨ {
    background: lime
}
.üí© {
    background: lime
}
.\\\\? {
    background: lime
}
.\\\\@ {
    background: lime
}
.\\\\. {
    background: lime
}
.\\\\3A \\\\) {
    background: lime
}
.\\\\3A \\\\\`\\\\( {
    background: lime
}
.\\\\31 23 {
    background: lime
}
.\\\\31 a2b3c {
    background: lime
}
.\\\\<p\\\\> {
    background: lime
}
.\\\\<\\\\>\\\\<\\\\<\\\\<\\\\>\\\\>\\\\<\\\\> {
    background: lime
}
.\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\[\\\\>\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\>\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\>\\\\+\\\\+\\\\+\\\\>\\\\+\\\\<\\\\<\\\\<\\\\<\\\\-\\\\]\\\\>\\\\+\\\\+\\\\.\\\\>\\\\+\\\\.\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\.\\\\.\\\\+\\\\+\\\\+\\\\.\\\\>\\\\+\\\\+\\\\.\\\\<\\\\<\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\+\\\\.\\\\>\\\\.\\\\+\\\\+\\\\+\\\\.\\\\-\\\\-\\\\-\\\\-\\\\-\\\\-\\\\.\\\\-\\\\-\\\\-\\\\-\\\\-\\\\-\\\\-\\\\-\\\\.\\\\>\\\\+\\\\.\\\\>\\\\. {
    background: lime
}
.\\\\# {
    background: lime
}
.\\\\#\\\\# {
    background: lime
}
.\\\\#\\\\.\\\\#\\\\.\\\\# {
    background: lime
}
.\\\\_ {
    background: lime
}
.\\\\{\\\\} {
    background: lime
}
.\\\\#fake\\\\-id {
    background: lime
}
.foo\\\\.bar {
    background: lime
}
.\\\\3A hover {
    background: lime
}
.\\\\3A hover\\\\3A focus\\\\3A active {
    background: lime
}
.\\\\[attr\\\\=value\\\\] {
    background: lime
}
.f\\\\/o\\\\/o {
    background: lime
}
.f\\\\\\\\o\\\\\\\\o {
    background: lime
}
.f\\\\*o\\\\*o {
    background: lime
}
.f\\\\!o\\\\!o {
    background: lime
}
.f\\\\'o\\\\'o {
    background: lime
}
.f\\\\~o\\\\~o {
    background: lime
}
.f\\\\+o\\\\+o {
    background: lime
}"
`;

exports[`/processor.js functionality values should support local values in value composition 1`] = `
Object {
  "fooga": Array [
    "fooga",
  ],
}
`;
